src/main/java/org/springframework/data/relational/core/conversion/RelationalEntityWriter.java

```diff
@@ -15,209 +15,240 @@
-			);
-		} else {
+		}
 
-			RelationalPersistentEntity<?> entity = context.getRequiredPersistentEntity(type);
-			IdentifierAccessor identifierAccessor = entity.getIdentifierAccessor(aggregateRoot);
+		private void write() {
 
-			deleteReferencedEntities(identifierAccessor.getRequiredIdentifier(), aggregateChange);
+			if (isNew(root)) {
 
-			Update<Object> update = DbAction.update(aggregateRoot, propertyPath, null);
-			aggregateChange.addAction(update);
+				setRootAction(new DbAction.InsertRoot<>(aggregateChange.getEntity()));
+				insertReferenced();
+			} else {
+				deleteReferenced();
 
-			referencedEntities(aggregateRoot).forEach(propertyAndValue -> insertReferencedEntities(propertyAndValue,
-					aggregateChange, propertyPath.nested(propertyAndValue.property.getName()), update));
+				setRootAction(new DbAction.UpdateRoot<>(aggregateChange.getEntity()));
+				insertReferenced();
+			}
 		}
-	}
 
-	private void insertReferencedEntities(PropertyAndValue propertyAndValue, AggregateChange aggregateChange,
-			RelationalPropertyPath propertyPath, DbAction dependingOn) {
+		//// Operations on all paths
 
-		Insert<Object> insert;
-		if (propertyAndValue.property.isQualified()) {
+		private void insertReferenced() {
+
+			paths.forEach(this::insertAll);
 
-			KeyValue valueAsEntry = (KeyValue) propertyAndValue.value;
-			insert = DbAction.insert(valueAsEntry.getValue(), propertyPath, dependingOn);
-			insert.getAdditionalValues().put(propertyAndValue.property.getKeyColumn(), valueAsEntry.getKey());
-		} else {
-			insert = DbAction.insert(propertyAndValue.value, propertyPath, dependingOn);
 		}
 
-		aggregateChange.addAction(insert);
-		referencedEntities(insert.getEntity()) //
-				.forEach(pav -> insertReferencedEntities( //
-						pav, //
-						aggregateChange, //
-						propertyPath.nested(pav.property.getName()), //
-						dependingOn) //
-		);
-	}
+		private void insertAll(PersistentPropertyPath<RelationalPersistentProperty> path) {
 
-	private Stream<PropertyAndValue> referencedEntities(Object o) {
+			from(path).forEach(node -> {
 
-		RelationalPersistentEntity<?> persistentEntity = context.getRequiredPersistentEntity(o.getClass());
+				DbAction.Insert<Object> insert;
+				if (node.path.getRequiredLeafProperty().isQualified()) {
 
-		return StreamUtils.createStreamFromIterator(persistentEntity.iterator()) //
-				.filter(PersistentProperty::isEntity) //
-				.flatMap( //
-						p -> referencedEntity(p, persistentEntity.getPropertyAccessor(o)) //
-								.map(e -> new PropertyAndValue(p, e)) //
-		);
-	}
+					KeyValue value = (KeyValue) node.getValue();
+					insert = new DbAction.Insert<>(value.value, path, getAction(node.parent));
+					insert.getAdditionalValues().put(node.path.getRequiredLeafProperty().getKeyColumn(), value.key);
+
+				} else {
+					insert = new DbAction.Insert<>(node.getValue(), path, getAction(node.parent));
+				}
+
+				previousActions.put(node, insert);
+				aggregateChange.addAction(insert);
```

Comment:
`AggregateChange` is mutated in deeply nested methods. We should rather return a collection of `DbAction<T>` and assemble `AggregateChange` in the top-most method.